src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -196,13 +249,26 @@ private async Task ProcessQueueAsync()
                         // will be sent back to the client but they can also be captured via HandleNonMutatingRequestError,
                         // though these errors don't put us into a bad state as far as the rest of the queue goes.
                         // Furthermore we use Task.Run here to protect ourselves against synchronous execution of work
-                        // blocking the request queue for longer periods of time (it enforces parallelizabilty).
-                        _ = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+                        // blocking the request queue for longer periods of time (it enforces parallelizability).
+                        var pendingTask = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+
+                        if (CancelInProgressWorkUponMutatingRequest)
+                        {
+                            pendingTasks.TryAdd(pendingTask, cancellationTokenSource);
+
+                            _ = pendingTask.ContinueWith(t =>
```

Comment:
I think the default behavior is to continue even if there is an exception, but its definitely something we should test.  For example the pending task throws for some reason, we would want to still run the continuation.

Reply:
It definitely continues if there is a cancellation requested, I'll verify with other types of exceptions, but this is very similar to how ExecuteAsync disposes of combinedTokenSource.